examples/vision/supervised-contrastive-learning.py

```diff
@@ -0,0 +1,186 @@
+
+# Convert the shape of the y_train and y_test from (50000, 1) to (50000, )
+y_train, y_test = tf.squeeze(y_train), tf.squeeze(y_test)
+
+# Scale images to the [0, 1] range
+x_train = x_train.astype("float32") / 255
+x_test = x_test.astype("float32") / 255
+
+print("x_train shape:", x_train.shape)
+print("y_train shape:", y_train.shape)
+print(x_train.shape[0], "train samples")
+print(x_test.shape[0], "test samples")
+
+"""## Build the encoder model
+
+The encoder model takes the image as an input and produce a 128-dimension feature vector.
+"""
+
+def create_encoder():
+  inputs = keras.Input(shape=input_shape)
+
+  x = layers.Conv2D(32, 3)(inputs)
+  x = layers.BatchNormalization()(x)
+  x = layers.Activation("relu")(x)
+  x = layers.Conv2D(64, 3)(x)
+  x = layers.BatchNormalization()(x)
+  x = layers.Activation("relu")(x)
+  x = layers.MaxPooling2D(pool_size=2)(x)
+
+  x = layers.Conv2D(128, 3)(x)
+  x = layers.BatchNormalization()(x)
+  x = layers.Activation("relu")(x)
+  x = layers.Conv2D(256, 3)(x)
+  x = layers.BatchNormalization()(x)
+  x = layers.Activation("relu")(x)
+  x = layers.MaxPooling2D(pool_size=2)(x)
+
+  outputs = layers.GlobalAveragePooling2D()(x)
+
+  model = keras.Model(inputs=inputs, outputs=outputs, name="cifar10-encoder")
+  return model
+
+encoder = create_encoder()
+encoder.summary()
+
+batch_size = 256
+num_epochs = 25
+dropout_rate = 0.5
+temperature = 0.05
+
+"""## Build the classification model
+
+The classification model adds a fully-connected layer on top of the encoder, plus a softmax layer with the target classes.
+"""
+
+def create_classifier(encoder, trainable=True):
+
+  for layer in encoder.layers:
+    layer.trainable = trainable
+
+  inputs = tf.keras.Input(shape=input_shape)
+  features = encoder(inputs)
+  features = tf.keras.layers.Dropout(dropout_rate)(features)
+  features = tf.keras.layers.Dense(64)(features)
+  features = tf.keras.layers.Dropout(dropout_rate)(features)
+  outputs =  tf.keras.layers.Dense(num_classes, activation="softmax")(features)
+  model = tf.keras.Model(inputs=inputs, outputs=outputs, name="cifar10-classifier")
+  model.compile(
+    optimizer=tf.keras.optimizers.Adam(),
+    loss=tf.keras.losses.SparseCategoricalCrossentropy(),
+    metrics=tf.keras.metrics.SparseCategoricalAccuracy()
+  )
+  return model
+
+"""## Experiment 1: Train the baseline classification model 
+
+In this experiment, a baseline classifier is trained normally, i.e., the encoder and the classifier parts are trained together as a single model to minimize cross-entropy loss.
+"""
+
+encoder = create_encoder()
+classifier = create_classifier(encoder)
+classifier.summary()
+
+history = classifier.fit(
+    x=x_train, 
+    y=y_train, 
+    batch_size=batch_size, 
+    epochs=num_epochs, 
+    validation_split=0.1)
+
+accuracy = classifier.evaluate(x_test, y_test)[1]
+print(f'Test accuracy: {round(accuracy*100, 2)}%')
+
+"""We get to ~69.8% validation accuracy.
```

Comment:
I'm still concerned that this is a very low score. We need to improve the model. Maybe you could try using a ResNet like architecture. The baseline score should be at least 80-85%. You could also try using KerasTuner to find a good model configuration.

70% accuracy can be done without even using deep learning btw.

(also, this should be "test accuracy", not "validation accuracy")

Reply:
Hi @fchollet - I think the purpose is to show the benefit of Supervised Constative Learning (SCL) over the conventional learning technique, rather than producing state of the art results for the cifar10. I think the model is more complex and produced better results in the conventional (baseline) model, it will produce even better results in the  SCL approach.

Also note that the model, in both cases, is only trained for 25 epochs.

If you agree with the rationale, I can add a sentence in the intro explaining this.

Reply:
CIFAR10 state of the art is 99%. I am not suggesting you should aim for the state of the art. I suggest getting "decent" results, which for a deep learning model is around 80-85%.

70% is far too low because it can be achieved without a convnet.

Reply:
@ksalama what do you think?

Reply:
@fchollet Ok - I used data augmentation and keras.applications.ResNet50V2 (random wrights)
- with 50 epochs, the baseline model achieves ~76% test acc, compared to ~81% with the SCL
- with 100 epochs, the baseline model achieves ~80 test acc, compared to ~82% with the SCL

It seems that ResNet50V2 is a bit big for cifar10 and tends to overfit the training dataset. However, code-wise, it is clean to use it as the encoder. In addition, ResNet50 was the model used in the research paper. I saw that a 90% test acc can be reach with data augmentation and a right model architecture[1]. However, as I mentioned, I wanted to illustrate how easy it is to implement SCL, and how it boost the performance of the model even with small number of epochs. I think SCL pays-off with "harder" datasets that required more complex models.
 
What are your thoughts? Do you think these experiments are good enough? If not, I can try the following:
1- Implement the architecture in [1]
2- Use another dataset (maybe cifar10) in which SCL can payoff more.

[1] https://github.com/moritzhambach/Image-Augmentation-in-Keras-CIFAR-10-/blob/master/CNN%20with%20Image%20Augmentation%20(CIFAR10).ipynb

Reply:
> It seems that ResNet50V2 is a bit big for cifar10 and tends to overfit the training dataset

Using this type of architecture is a good idea. You could simplify the model a bit (fewer blocks, smaller blocks) to right-size it for CIFAR10 (this architecture was developed for ImageNet which is much larger).

Here's a simple architecture you could use, a kind of mini-xception model: https://colab.research.google.com/drive/1qeFE_xkC1qnbbrJkmAw3hRFspu-d_OVn?usp=sharing

(you can tune the number of blocks and their size easily, e.g. with KerasTuner)

Reply:
Thank you for sharing the mini-xception notebook! 

In this example, I think it would be nice to show how to use one of the Keras application models as-is to be the encoder, especially as the focus of the tutorial is not to show how to build a powerful vision architecture, rather how SCL can improve the performance of "a given" architecture. Thus, using an out-of-the-box model (like ResNet50V2) will make the focus of the tutorial on the SCL technique, rather than the details of the fancy architecture of mini-xception :-)

So my preference is use the current implementation for the tutorial. However, if you think it is necessary for tutorial to be published, I am happy to try this mini-xception model.

What do you think?